### PR TITLE
feat: add cmake lib for arducam mega

### PR DESCRIPTION
**Pull request by anujdeshpande**

### CODE REVIEW

**Comment by Bucknalla** [comment]
I think your header file has an extra '.' after the file name

File: arducam_mega.c [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+	.slave = 0,
+};
+
+#define SPI_DEV_NAME DEVICE_DT_NAME(DT_NODELABEL(spi0))
+const struct device *spi;
+
+#define LED DT_ALIAS(csled)
+struct gpio_dt_spec spec = GPIO_DT_SPEC_GET(LED, gpios);
+
+uint8_t cameraBusRead(uint8_t address)
```
Comment (Bucknalla) [inline arducam_mega.c]: camera_bus_read

File: arducam_mega.c [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+	};
+	struct spi_buf_set rx_bufs = {.buffers = rx_buf, .count = 2};
+
+	ret = spi_read(spi, &spi_cfg, &rx_bufs);
+	gpio_pin_toggle_dt(&spec);
+	spi_release(spi, &spi_cfg);
+	return rxdata[1];
+}
+
+uint8_t cameraReadReg(uint8_t addr)
```
Comment (Bucknalla) [inline arducam_mega.c]: camera_read_reg

File: arducam_mega.c [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+	return temp;
+}
+
+#define BUFFER_SIZE 0xff
+uint8_t imageData = 0;
+uint8_t imageDataNext = 0;
+uint8_t headFlag = 0;
+unsigned int i = 0;
+uint8_t imageBuff[BUFFER_SIZE] = {0};
+int received_length = 0;
```
Comment (Bucknalla) [inline arducam_mega.c]: Can we better organise this and change the variable scope from global to their corresponding function?

File: arducam_mega.c [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+#include <zephyr/syscall_handler.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/device.h>
+#include <zephyr/fs/fs.h>
+
+#define LOG_MODULE_NAME arducam_mega
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#define MAX_PATH          51200
+#define SOME_REQUIRED_LEN 30
```
Comment (Bucknalla) [inline arducam_mega.c]: Better name for this define, something with a bit more clarity?

File: arducam_mega.c [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+		LOG_ERR("Not enough concatenation buffer to create file paths");
+		return;
+	}
+
+	strncpy(path, base_path, sizeof(path));
+
+	path[base++] = '/';
+	path[base] = 0;
+	strcat(&path[base], filename);
+	int ret;
```
Comment (Bucknalla) [inline arducam_mega.c]: Move to top of function

File: arducam_mega.c [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+		}
+		if (imageData == 0xff && imageDataNext == 0xd8 && file_opened == 0) {
+			ret = fs_open(&file, path, FS_O_CREATE | FS_O_WRITE | FS_O_APPEND);
+			if (ret != 0) {
+				LOG_ERR("Failed to create file %s %d", path, ret);
+				return;
+			}
+			LOG_INF("Opened file successfully\n");
+			file_opened = 1;
+			k_sleep(K_MSEC(3000));
```
Comment (Bucknalla) [inline arducam_mega.c]: Why is this neccessary?

File: arducam_mega.c [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+	gpio_pin_toggle_dt(&spec);
+	k_sleep(K_MSEC(10));
+	return 1;
+}
+void cameraWriteReg(uint8_t addr, uint8_t val)
+{
+	cameraBusWrite(addr | 0x80, val);
+}
+
+void camera_wait_idle()
```
Comment (Bucknalla) [inline arducam_mega.c:83]: Can we do this a better way? Or is the only way for the camera to tell us that it is idle, is by setting its register?

File: arducam_mega.c [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+	LOG_INF("Initializing the camera");
+	spi = device_get_binding(SPI_DEV_NAME);
+	if (!device_is_ready(spi)) {
+		LOG_ERR("Device SPI not ready, aborting test");
+		return -1;
+	}
+	gpio_pin_toggle_dt(&spec);
+	k_sleep(K_MSEC(1000));
+	gpio_pin_toggle_dt(&spec);
+	k_sleep(K_MSEC(1000));
```
Comment (Bucknalla) [inline arducam_mega.c]: I assume this is what Arducam are doing?

**Review by Bucknalla — CHANGES_REQUESTED** [review]
I'm being a bit more picky here as this potentially needs to be a good example driver for other projects to use. Using https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format for code formatting.

**Review by Bucknalla — CHANGES_REQUESTED** [review]
I'm being a bit more picky here as this potentially needs to be a good example driver for other projects to use. Using https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format for code formatting.

File: arducam_mega.c (hunk 6 above)
Comment (anujdeshpande) [reply arducam_mega.c]: It isn't - it's from when I was using it for testing something. Removed!

File: arducam_mega.c (hunk 7 above)
Comment (anujdeshpande) [reply arducam_mega.c:83]: We keep reading a reg till it has a specific value - which indicates that we can continue to the next operation. Arducam's lib doesn't have a different way of doing this - perhaps when they get us the datasheet we can revisit this

**Pull request closed without merging**